packages/nimble-components/src/table/models/keyboard-navigation-manager.ts

```diff
@@ -0,0 +1,1345 @@
+        );
+        this.table.viewport.removeEventListener(
+            'cell-focus-in',
+            this.onCellFocusIn as EventListener
+        );
+        this.table.viewport.removeEventListener(
+            'cell-blur',
+            this.onCellBlur as EventListener
+        );
+    }
+
+    public handleChange(source: unknown, args: unknown): void {
+        let focusRowAndCell = false;
+        if (source === this.virtualizer && args === 'visibleItems') {
+            focusRowAndCell = true;
+        } else if (source === this.table && args === 'rowElements') {
+            for (const notifier of this.visibleRowNotifiers) {
+                notifier.unsubscribe(this);
+            }
+            this.visibleRowNotifiers = [];
+            for (const visibleRow of this.table.rowElements) {
+                const rowNotifier = Observable.getNotifier(visibleRow);
+                rowNotifier.subscribe(this, 'rowStateIndex');
+                if (visibleRow.rowStateIndex === this.rowIndex) {
+                    focusRowAndCell = true;
+                }
+            }
+        } else if (args === 'rowStateIndex') {
+            const index = (source as TableRow | TableGroupRow).rowStateIndex;
+            if (index === this.rowIndex) {
+                focusRowAndCell = true;
+            }
+        }
+
+        if (focusRowAndCell) {
+            // Focusable elements in cells, and action menus, are both blurred on scroll. To maintain our row/cell focus state,
+            // we focus the cell instead here. (We also don't want to refocus the cell content when the focusedRecycleCallback just
+            // blurred it.)
+            if (
+                this.focusType === TableFocusType.cellActionMenu
+                || this.focusType === TableFocusType.cellContent
+            ) {
+                this.setCellFocusState(this.columnIndex, this.rowIndex, false);
+            }
+            if (this.inNavigationMode && this.hasRowOrCellFocusType()) {
+                if (this.rowIndex > this.table.tableData.length - 1) {
+                    // Focused row index no longer valid, coerce to 1st row if possible
+                    if (this.table.tableData.length > 0) {
+                        this.rowIndex = 0;
+                    } else {
+                        if (this.focusWithinTable) {
+                            this.setDefaultFocus();
+                        } else {
+                            this.focusType = TableFocusType.none;
+                        }
+                        return;
+                    }
+                }
+                if (this.focusWithinTable) {
+                    this.focusCurrentRow(false);
+                }
+            }
+        }
+    }
+
+    public onRowFocusIn(event: FocusEvent): void {
+        if (this.isCurrentlyFocusingElement) {
+            return;
+        }
+        const row = event.target;
+        if (row instanceof TableRow || row instanceof TableGroupRow) {
```

Comment:
Once we have a name for this set of rows, i.e. flattenedRows, resolvedRows, visualRows we can make a type predicate helper for this check

```ts
    private isVisualRow(row: unknown): row is TableRow | TableGroupRow {
        return row instanceof TableRow || row instanceof TableGroupRow;
    }
```

Reply:
Updated